From bf5ea0cbfaa64a554695d71999df3b54ab6f6734 Mon Sep 17 00:00:00 2001
From: Howard Hsu <howard-yh.hsu@mediatek.com>
Date: Fri, 14 Jun 2024 14:53:22 +0800
Subject: [PATCH 118/137] mtk: hostapd: support ap_wireless vendor cmd with
 link id

Support ap_wireless vendor cmd including link id in nl80211 msg.

Signed-off-by: Howard Hsu <howard-yh.hsu@mediatek.com>
---
 src/ap/ap_drv_ops.c          | 9 ++++++++-
 src/common/mtk_vendor.h      | 2 ++
 src/drivers/driver.h         | 3 ++-
 src/drivers/driver_nl80211.c | 6 +++++-
 4 files changed, 17 insertions(+), 3 deletions(-)

diff --git a/src/ap/ap_drv_ops.c b/src/ap/ap_drv_ops.c
index ae2be90b6..8631bf960 100644
--- a/src/ap/ap_drv_ops.c
+++ b/src/ap/ap_drv_ops.c
@@ -1377,9 +1377,16 @@ int hostapd_drv_get_aval_bss_color_bmp(struct hostapd_data *hapd, u64 *aval_colo
 
 int hostapd_drv_ap_wireless(struct hostapd_data *hapd, u8 sub_vendor_id, int value)
 {
+	s8 link_id = -1;
+
 	if (!hapd->driver || !hapd->driver->ap_wireless)
 		return 0;
-	return hapd->driver->ap_wireless(hapd->drv_priv, sub_vendor_id, value);
+
+	if (hapd->conf->mld_ap)
+		link_id = hapd->mld_link_id;
+
+	return hapd->driver->ap_wireless(hapd->drv_priv, sub_vendor_id, value,
+					 link_id);
 }
 
 int hostapd_drv_ap_rfeatures(struct hostapd_data *hapd, u8 sub_vendor_id, int value)
diff --git a/src/common/mtk_vendor.h b/src/common/mtk_vendor.h
index c6de8862b..937b968d5 100644
--- a/src/common/mtk_vendor.h
+++ b/src/common/mtk_vendor.h
@@ -165,6 +165,8 @@ enum mtk_vendor_attr_wireless_ctrl {
 	MTK_VENDOR_ATTR_WIRELESS_CTRL_AMSDU,
 	MTK_VENDOR_ATTR_WIRELESS_CTRL_CERT = 9,
 	MTK_VENDOR_ATTR_WIRELESS_CTRL_RTS_SIGTA,
+	MTK_VENDOR_ATTR_WIRELESS_CTRL_MU_EDCA,
+	MTK_VENDOR_ATTR_WIRELESS_CTRL_LINK_ID,
 
 	/* keep last */
 	NUM_MTK_VENDOR_ATTRS_WIRELESS_CTRL,
diff --git a/src/drivers/driver.h b/src/drivers/driver.h
index aa158aa05..dacf0a98d 100644
--- a/src/drivers/driver.h
+++ b/src/drivers/driver.h
@@ -5334,8 +5334,9 @@ struct wpa_driver_ops {
 	* ap_wireless - set wireless command
 	* @priv: Private driver interface data
 	* @value: value
+	* @link_id: MLD link id. -1 if this is an non-MLD AP.
 	*/
-	int (*ap_wireless)(void *priv, u8 mode, int value);
+	int (*ap_wireless)(void *priv, u8 mode, int value, s8 link_id);
 
 	/**
 	* ap_rfeatures - set ap rf features command
diff --git a/src/drivers/driver_nl80211.c b/src/drivers/driver_nl80211.c
index edda40c48..80fe2e591 100644
--- a/src/drivers/driver_nl80211.c
+++ b/src/drivers/driver_nl80211.c
@@ -141,6 +141,7 @@ wireless_ctrl_policy[NUM_MTK_VENDOR_ATTRS_WIRELESS_CTRL] = {
 	[MTK_VENDOR_ATTR_WIRELESS_CTRL_BA_BUFFER_SIZE] = {.type = NLA_U16 },
 	[MTK_VENDOR_ATTR_WIRELESS_CTRL_AMSDU] = {.type = NLA_U8 },
 	[MTK_VENDOR_ATTR_WIRELESS_CTRL_CERT] = {.type = NLA_U8 },
+	[MTK_VENDOR_ATTR_WIRELESS_CTRL_LINK_ID] = {.type = NLA_U8 },
 };
 
 static struct nla_policy
@@ -14910,7 +14911,7 @@ static int nl80211_get_aval_color_bmp(void *priv, u64 *aval_color_bmp)
 	return ret;
 }
 
-static int nl80211_ap_wireless(void *priv, u8 sub_vendor_id, int value)
+static int nl80211_ap_wireless(void *priv, u8 sub_vendor_id, int value, s8 link_id)
 {
 	struct i802_bss *bss = priv;
 	struct wpa_driver_nl80211_data *drv = bss->drv;
@@ -14941,6 +14942,9 @@ static int nl80211_ap_wireless(void *priv, u8 sub_vendor_id, int value)
 	else
 		nla_put_u8(msg, sub_vendor_id, (u8) value);
 
+	if (link_id > -1)
+		nla_put_u8(msg, MTK_VENDOR_ATTR_WIRELESS_CTRL_LINK_ID, link_id);
+
 	nla_nest_end(msg, data);
 	ret = send_and_recv_cmd(drv, msg);
 	if (ret)
-- 
2.18.0

